um_run.py

```diff
@@ -0,0 +1,315 @@
+        (['.sh'], False),
+    ]
+
+    return Config(
+        label=project_name,
+        workspace=workspace,
+
+        # run params
+        # use_multiprocessing=False,
+        debug_skip=True,
+
+        # source config - this will not be part of build, we have existing code to look at for this
+        grab_config=grab_config,
+
+        # build steps
+        steps=[
+
+            MyCustomCodeFixes(name="my custom code fixes"),
+
+            FindSourceFiles(source_root=workspace / SOURCE_ROOT, file_filtering=file_filtering),  # template?
+
+            RootIncFiles(workspace / SOURCE_ROOT),
+
+            CPragmaInjector(),
+
+            CPreProcessor(
+                source=Artefact('pragmad_c'),
+                preprocessor='cpp',
+                path_flags=[
+                    # todo: this is a bit "codey" - can we safely give longer strings and split later?
+                    AddFlags(match="$source/um/*", flags=[
+                        '-I', '$source/um/include/other',
+                        '-I', '$source/shumlib/common/src',
+                        '-I', '$source/shumlib/shum_thread_utils/src']),
+
+                    AddFlags(match="$source/shumlib/*", flags=[
+                        '-I', '$source/shumlib/common/src',
+                        '-I', '$source/shumlib/shum_thread_utils/src']),
+
+                    # todo: just 3 folders use this
+                    AddFlags("$source/um/*", ['-DC95_2A', '-I', '$source/shumlib/shum_byteswap/src']),
+                ],
+            ),
+
+            # todo: explain fnmatch
+            FortranPreProcessor(
+                preprocessor='cpp',
+                common_flags=['-traditional-cpp', '-P'],
+                path_flags=[
+                    AddFlags("$source/jules/*", ['-DUM_JULES']),
+                    AddFlags("$source/um/*", ['-I', '$relative/include']),
+
+                    # coupling defines
+                    AddFlags("$source/um/control/timer/*", ['-DC97_3A']),
+                    AddFlags("$source/um/io_services/client/stash/*", ['-DC96_1C']),
+                ],
+            ),
+
+            Analyse(
+                root_symbol='um_main',
+                unreferenced_deps=None,
+
+                # fparser2 fails to parse this file, but it does compile.
+                special_measure_analysis_results=[
+                    AnalysedFile(
+                        fpath=Path(
+                            os.path.expanduser("~/git/fab/tmp-workspace/um_atmos_safe/build_output/casim/lookup.f90")),
+                        file_hash=None,
+                        symbol_defs=['lookup'],
+                        symbol_deps=['mphys_die', 'variable_precision', 'mphys_switches', 'mphys_parameters', 'special',
+                                     'passive_fields', 'casim_moments_mod', 'yomhook', 'parkind1'],
+                        file_deps=[],
+                        mo_commented_file_deps=[]),
+                ]
```

Comment:
Agreed. A future speed optimisation might see the use of fparser limited to reading lines rather than full parsing, which might also make this problem go away.